### DISCUSSION TITLE
Product save and update doesnt do the same thing.

### DISCUSSION THREAD

**Discussion by Ifnir** [getcandy/candy-api#394] [Ideas]
When I run

```
$product = new Product();

$attributes = [];

$attributes['name']['da'] = "Name";
$attributes['page_title']['da'] = "A Title";
$attributes['short_description']['da'] = "Text1";
$attributes['description']['da'] = "Text1";
$attributes['erp_id']['da'] = "asdasd-sadasd-asdasdasd-asdasd";

$product->attribute_data = $attributes;
$product->option_data = [];

$product->save();
```

I get this in my MySQL column, which is what I want it to be.

`{"name": {"webstore": {"da": "Name", "en": null}}, "erp_id": {"webstore": {"da": "asdasd-sadasd-asdasdasd-asdasd", "en": null}}, "page_title": {"webstore": {"da": "A Title", "en": null}}, "description": {"webstore": {"da": "Text1", "en": null}}, "short_description": {"webstore": {"da": "Text1", "en": null}}}`

which is what I want, but when I run this code:

```
$attributes = [];

$attributes['name']['da'] = "Name";
$attributes['page_title']['da'] = "A Title";
$attributes['short_description']['da'] = "Text2";
$attributes['description']['da'] = "Text2";
$attributes['erp_id']['da'] = "asdasd-sadasd-asdasdasd-asdasd";

$product->attribute_data = $attributes;
$product->update();
```

Then is get this in my MySQL column

`{"name": {"da": "Name"}, "erp_id": {"da": "asdasd-sadasd-asdasdasd-asdasd"}, "page_title": {"da": "A Title"}, "description": {"da": "Text2"}, "short_description": {"da": "Text2"}}
`
I want to run clean code as possible and avoid making a separate function to update a product or a variant for that matter.
Is it possible in the model to change to attribute setter to also do it when its update?

**Comment by glennjacobs** [comment]
Are you doing this to bulk import/update products?

**Reply by glennjacobs** [reply]
I can see what is happening here and I think this topic has been raised before. It's all related to

```php
    public function setAttributeDataAttribute($val)
    {
        if (! $this->id) {
            $this->attributes['attribute_data'] = json_encode($this->mapAttributes($val));
        } else {
            // dd(json_encode($val));
            $this->attributes['attribute_data'] = json_encode($val);
        }
    }
```

On initial set-up it takes one form, but it assumes updates will come back in the correct JSON format from the API, so it doesn't need to re-map it.

For the moment, you will have to map it differently for updates. I'm going to raise and discuss this internally, as it seems a bit wrong.

**Reply by Ifnir** [reply]
Yeah

It works if I put
$attributes['short_description']['webstore]['da'] = "Text2";
but that doesnt work. when I import a new product.

**Reply by glennjacobs** [reply]
I'm thinking the `$this->mapAttributes()` function should be called separately outside of the setter function.

**Reply by glennjacobs** [reply]
I've mocked up a draft idea at #395, needs plenty of testing before we could merge, however.

It would mean you could create and update product attribute data in the same format, but you would need to include the channel.
